### ISSUE TITLE
Follow up to #2917 "Memory leak in OpenCL backend"

### ISSUE THREAD

**Issue by laser101** [arrayfire/arrayfire#3006]
Proposal to provide an interim workaround to safely delete cl_mem* pointers returned by `array::device()` until this potential memory leak can be addressed properly (as suggested in #2917).

Issue #2917 extended the OpenCL interop documentation to advise users to properly delete cl_mem* pointers, but this is not straightforward on MS Windows if the user's application is built with a different C++ runtime library than the released ArrayFire library.
 
For example, the [Interoperability with OpenCL](http://arrayfire.org/docs/interop_opencl.htm) documentation includes an example "Adding custom OpenCL kernels to an existing ArrayFire application". If this application is compiled with the normal MS Visual Studio debug options and linked against the ArrayFire release build, then it will throw an exception when executing the delete statements...
```c++
// ...

// 3. Obtain cl_mem references to af::array objects
cl_mem* d_A = A.device<cl_mem>();
cl_mem* d_B = B.device<cl_mem>();

// ...

// Delete the pointers returned by the device function. This does NOT
// delete the cl_mem memory and only deletes the pointers
delete d_A;
delete d_B;
```

This exception is thrown because `d_A` and `d_B` are allocated by the Arrayfire library which is built against the release runtime libraries, whereas the application has been compiled against the debug runtime. Now there are ways to overcome this based on forcing MSVS to compile using the release runtime even when producing a debug build, but it gets complicated for large projects.

The proposal here is to add an interim workaround in the ArrayFire library to address this issue...
```c++
template <typename T>
void afcl::_delete<T>(T* ptr) {
    delete ptr;
}
```
then instantiate for `cl_mem*` (and any other types required) to export in the `af.dll`.

Alternatively specifically handle the `cl_mem*` case...
```c++
void afcl::_delete(cl_mem* ptr) {
    delete ptr;
}
```
In this way, the delete operation is performed within the ArrayFire library, and is therefore executed by the same C++ runtime library as used to allocate the memory, thus avoiding the exeception regardless of whether the application is linked against the debug or release runtime libraries.

I've prefixed the proposed function name with an underscore to indicate that this is a "hidden" or "interim" implementation, but other naming options obviously exist.

I should point out that I have not yet been able to test this proposal, but interested in feedback before I invest further time.

**Comment by 9prady9** [comment]
@laser101 I think the solution suggested by @umar456 [here](https://github.com/arrayfire/arrayfire/issues/2917#issuecomment-636993181) about adding another version of API suffixed by V2 is good way to go about it I think. We are due a 3.8 release soon and hence we can introduce this new V2 version of this array class method into that. @umar456 What do you think ?

**Comment by laser101** [comment]
@9prady9 I agree that the suggestion by @umar456 is a better approach and support its inclusion asap. Since the timeline wasn't clear, my proposal above was intended as a simple interim workaround until @umar456 approach could be implemented.

It would be great if the v2 version could be included in the 3.8 release.
